R/undesirable_function_linter.R

```diff
@@ -51,17 +57,32 @@
 #'   linters = undesirable_function_linter(fun = c("dir" = NA), symbol_is_undesirable = FALSE)
 #' )
 #'
+#' lint(
+#'   text = 'dir <- "path/to/a/directory"',
+#'   linters = undesirable_function_linter(fun = "dir", symbol_is_undesirable = FALSE)
+#' )
+#'
 #' @evalRd rd_tags("undesirable_function_linter")
 #' @seealso [linters] for a complete list of linters available in lintr.
 #' @export
 undesirable_function_linter <- function(fun = default_undesirable_functions,
                                         symbol_is_undesirable = TRUE) {
-  stopifnot(is.logical(symbol_is_undesirable))
-  if (is.null(names(fun)) || !all(nzchar(names(fun))) || length(fun) == 0L) {
-    cli_abort(c(
-      x = "{.arg fun} should be a non-empty named character vector.",
-      i = "Use missing elements to indicate default messages."
-    ))
+  if (is.list(fun)) fun <- unlist(fun)
+  stopifnot(
+    is.logical(symbol_is_undesirable),
+    # allow (uncoerced->implicitly logical) 'NA'
+    `\`fun\` should be a non-empty character vector` =
+      length(fun) > 0L && (is.character(fun) || all(is.na(fun)))
+  )
+
+  nm <- names2(fun)
+  implicit_idx <- !nzchar(nm)
+  if (any(implicit_idx)) {
+    names(fun)[implicit_idx] <- fun[implicit_idx]
+    is.na(fun) <- implicit_idx
+  }
+  if (anyNA(names(fun))) {
+    cli_abort("Found missing entries to {.arg fun}: {.val {which(is.na(names(fun)))}}")
```

Comment:
The message could be worded more actionably.

Maybe something like "Elements of op must be named or non-empty. Index ... is empty."?

Reply:
PTAL